c10/core/TensorImpl.h

```diff
@@ -1573,6 +1574,42 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     }
   }
 
+  /**
+   * Faster implementation of refresh_contiguous() that can be used if
+   * we know the current MemoryFormat.
+   */
+  void refresh_contiguous(MemoryFormat memory_format) {
+    // NOTE: Make sure to keep the other overload in sync with this implementation!
+    is_contiguous_ = memory_format == MemoryFormat::Contiguous || compute_contiguous();
+    switch (memory_format) {
+      case MemoryFormat::Contiguous:
+        is_channels_last_contiguous_ = false;
+        is_channels_last_contiguous_ = false;
+        is_channels_last_3d_contiguous_ = false;
```

Comment:
`is_coniguous_` and `is_channels_last_contiguous_` are not mutually exclusive, e.g. contiguous tensor of size (N,C,1,1) has both as `true`.

Reply:
is there anything useful about this diff, then? I don't understand all the subtleties here.

Reply:
There are useful things, contiguous computations are repeated twice now, and they shouldn't be. But you are right, it's more annoying. I should still do it.